### ISSUE TITLE
Never Ending Reddit Doesn't Automatically Load and NOT Paused

### ISSUE THREAD

**Issue by GuardianMajor** [honestbleeps/Reddit-Enhancement-Suite#4781]
- My browser Firefox 57 on Windows 10
- My RES version: 5.12.4
- Am I in reddit beta (yes/no): yes

*What's the problem?*

My Never Ending Reddit feature has stopped working, it will pause at the bottom of each page, even though I do not have the pause button checked, never have. You have to click to get it to load more, no longer happens automatically and I have not changed anything since it was working before. on the 5.11.x branch.

EDIT: Just encountered something else. After page 1, you click it will load page 2 but after that, even clicking it will no longer load any more pages. Just noticed that too.

**_My current setting:_**

![image](https://user-images.githubusercontent.com/5274181/40630013-df19747c-6283-11e8-810b-0fd5bbf7a8f9.png)

**_Current behavior:_**

![image](https://user-images.githubusercontent.com/5274181/40629988-b714a686-6283-11e8-84bd-f12e7b56a619.png)

*Are there any related posts in https://github.com/honestbleeps/Reddit-Enhancement-Suite/issues or /r/RESissues?*

I searched to see if it has been reported and I didn't see anything, apologies if I missed it somehow. The only thing I saw was a similar closed issue that said it was fixed in 5.4 but otherwise I didn't see anything about this. Am I doing something wrong with my settings somewhere? Thanks in advance.

*What other browser extensions are installed?*

uBlock, GreaseMonkey, ScreenGrab, and RES

**Comment by larsjohnsen** [comment]
Does it work with a clean profile? https://developer.mozilla.org/en-US/Firefox/Multiple_profiles#Creating_a_profile_through_the_Firefox_browser

**Comment by GuardianMajor** [comment]
@larsjohnsen Thanks for the link but I know how to do it, not a novice there. But honestly, I have not tried it and didn't see the point but if it helps then by all means, will do that. Not sure how a clean profile will differ for something like this but we'll find out I guess.

Just so I know what to look for, can you tell me why you think this is a profile issue when it is so specific to a single feature? I am curious to understand the thinking beyond just being a prophylactic measure.

**Comment by GuardianMajor** [comment]
For what it is worth, this is the error that shows up in the console:

> ReferenceError: IntersectionObserver is not defined  foreground.entry.js:14849:8

And it is arising from this code block (where the --> is):

```
function attachLoaderWidget(widget) {
	loaderWidget = widget;
	$(neverEndingReddit_siteTable()).after(widget);

-->	const prefetchIo = new IntersectionObserver(([{ isIntersecting }]) => {
		if (!isIntersecting) return;
		prefetchNextPage();
		prefetchIo.disconnect();
	}, { rootMargin: '100%' });
	prefetchIo.observe(widget);

	initAutoNextPage = once_default()(() => {
		const loadNextPageIo = new IntersectionObserver(([{ isIntersecting }]) => {
			if (!isIntersecting) return;
			loadNextPage();
			loadNextPageIo.disconnect();
		}, { threshold: [1] });
		loadNextPageIo.observe(widget);
	});

	if (neverEndingReddit_module.options.autoLoad.value && !isPaused && !pauseAfterPage()) initAutoNextPage();
}
```

**Comment by larsjohnsen** [comment]
That's weird. `IntersectionObserver` (which Never Ending Reddit is dependent on since v5.12.4) should be available in Firefox 57, so I don't know what causes this issue. The only suggestions I have are to try a clean profile in case your Firefox state is messed up, or upgrade to the latest version of Firefox.

**Comment by GuardianMajor** [comment]
Ok, then that's that. I have tried in a clean profile, same result. It won't trigger automatically and after the first time you click it and gives you page 2, it will no longer even respond to even clicks to get any more pages, just dies at the end of page 2.

It seems that it was introduced in 5.12.x which is when it stopped working for me, so that would make sense but not sure why still, that's the odd part. Anyway, thank you. I guess I need to either roll back to when it worked or "rig" the newer version with the old code block for this feature, we'll see.

**Comment by Alex9001** [comment]
Same here. Worked yesterday, won't work today :(

**Comment by larsjohnsen** [comment]
@Alex9001 You are experiencing a another issue not related to the one described here. See [the wiki](https://www.reddit.com/r/Enhancement/wiki/faq/never_ending_reddit) for how to fix it.

**Comment by GuardianMajor** [comment]
Ok, I wanted to share my findings. The reason it is not working has to do with this setting in Firefox:

`dom.IntersectionObserver.enabled`

If this is set to `false` which is the case in many setups that want to avoid the specialized use of it by ads and trackers, then this will make this feature fail. 

>  * Almost a year to complete, three versions late to stable (as default false),
>  * number 1 cause of crashes in nightly numerous times,
>  * and is (primarily) an ad network API for "ad viewability checks" down to a pixel level
>  1. https://developer.mozilla.org/docs/Web/API/Intersection_Observer_API
>  2. https://w3c.github.io/IntersectionObserver/
>  3. https://bugzilla.mozilla.org/1243846
>  4. https://bugzilla.mozilla.org/1381574

Given that the user has to choose between this feature or not being tracked by ads, can we find an alternative to using this or provide an option to NOT use `IntersectionObserver` for those who know it will break with this config set to false?

**Comment by larsjohnsen** [comment]
RES uses the API for many of the same reasons the ad networks find is useful: It is a lot cheaper than forcing a reflow to check whether something is in view. I find it little believable that disabling it prevents you from being tracked by ads.

**Comment by GuardianMajor** [comment]
I appreciate that it is easier and the now edited comment you made not withstanding, the goal of a project is to be as widely compatible as possible. No need to be dismissive and/or insulting when sometimes use of specialized functionality has to be weighed against that goal. 

You are probably right that it doesn't matter and I didn't say that it matters in an existential manner, but just pointing out that since most security tools tend to disable it, a great number of users will be affected, without knowing why. Specially that no one here had a clue why it wasn't working or offered an explanation as to why something they chose to use made a long standing feature stop working.

I had to find out on my own as no one else had a clue, so that's why I shared it so that at least now there is light on the subject in case someone else encounters it. How you use that information to be more inclusive is your call and taking a dismissive tone and closing this issue pretty much makes that position clear. Thanks.

**Comment by larsjohnsen** [comment]
I appreciate you bringing the configuration to attention. RES has been using IntersectionObserver in Firefox since it was enabled by default and this is the first report of RES breaking due it being overriden. Nevertheless, if the security measures you describe is shown to be widespread, then a workaround will certainly be considered.

**Comment by erikdesjardins** [comment]
I figured it was `dom.IntersectionObserver.enabled`, but it didn't work for you on a clean profile, where that preference is enabled. Unless you're using a custom build of Firefox?

> the user has to choose between this feature or not being tracked by ads

How inflammatory.

IntersectionObserver allows "tracking" that an ad was visible on screen, not tracking in the usual sense of the word (that is, fingerprinting a user to correlate them across IPs or domains). More importantly, it's a JavaScript API, so only ads which are unblocked and able to run JS can use it. If you care about privacy you should be blocking non-first-party and inline JS (I do), which effectively blocks IO as well. If you aren't blocking JS, blocking IO alone is odd: third-party ads won't be able to check if they're on screen, sure, but they can still phone home, fingerprint your browser through timing or related loopholes, and other worse things.

If ads are first-party, IO provides no additional information: pixel-level visibility is available through getBoundingClientRect, summing offsets, etc.

**Comment by GuardianMajor** [comment]
Yes, it did not work in a clean profile, and if you had said that was what you suspected, I could have actually checked it, but you didn't. A clean profile means nothing if there are policy in effect that are loaded by default, you need to perhaps understand that in the future if using the clean profile as meaning, everything is default.

It is not inflammatory at all, don't see how you are beating a dead horse into something it is not, but it is a statement of fact that it is indeed used in that vector. A kitchen knife is used for cooking but can also be used to kill someone, no doubt features don't have just a single use case, no one is arguing that.

Just because a feature exists and it makes it easy to do something, doesn't mean it is the best approach, just an approach. What I should, do already or don't do is not a matter of discussion or dictation here, as I am aware how an API is used and in what context - so stay on topic. 

Not all tracking is done via third-party which is by default disabled for me, as this is not my first day using the internet. But if you think that it only happens in third party situations, then I will return the same unsolicited advise back to you that you are taking a narrow and dismissive view lacking insight. No, without certain API being readily available inside the browser, there is less chance a site can get the information using purely scripting as it will have to work harder - exposing its methodology which can be neutered easily or won't get as granular of information as when it can hook the browser to get the whole kitchen sink provided to it.

Anyway, thanks for the input but I don't want to get too far off topic discussing philosophy or whatnot that is not relevant to the issue.
